osquery/events/linux/bpf/processcontextfactory.cpp

```diff
@@ -0,0 +1,279 @@
+/**
+ * Copyright (c) 2014-present, The osquery authors
+ *
+ * This source code is licensed as defined by the LICENSE file found in the
+ * root directory of this source tree.
+ *
+ * SPDX-License-Identifier: (Apache-2.0 OR GPL-2.0-only)
+ */
+
+#include <osquery/events/linux/bpf/processcontextfactory.h>
+#include <osquery/events/linux/bpf/uniquedir.h>
+
+#include <dirent.h>
+#include <fcntl.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+
+#ifndef O_PATH
+#define O_PATH 010000000
+#endif
+
+namespace osquery {
+namespace {
+const std::size_t kMaxFileSize{1024 * 100};
+const std::string kProcFsRoot{"/proc/"};
+} // namespace
+
+bool ProcessContextFactory::captureSingleProcess(
+    ProcessContext& process_context, pid_t process_id) const {
+  return captureSingleProcess(*fs.get(), process_context, process_id);
+}
+
+bool ProcessContextFactory::captureAllProcesses(
+    ProcessContextMap& process_map) const {
+  return captureAllProcesses(*fs.get(), process_map);
+}
+
+ProcessContextFactory::ProcessContextFactory(
+    IFilesystem::Ref filesystem_interface) {
+  fs = std::move(filesystem_interface);
+}
+
+bool ProcessContextFactory::captureSingleProcess(
+    IFilesystem& fs, ProcessContext& process_context, pid_t process_id) {
+  process_context = {};
+
+  tob::utils::UniqueFd process_root;
+  if (!fs.open(process_root,
+               kProcFsRoot + std::to_string(process_id),
+               O_DIRECTORY)) {
+    return false;
+  }
+
+  tob::utils::UniqueFd process_fdmap;
+  if (!fs.openAt(process_fdmap, process_root.get(), "fd", O_DIRECTORY)) {
+    return false;
+  }
+
+  tob::utils::UniqueFd process_cmdline;
+  if (!fs.openAt(process_cmdline, process_root.get(), "cmdline", O_RDONLY)) {
+    return false;
+  }
+
+  tob::utils::UniqueFd process_stat;
+  if (!fs.openAt(process_stat, process_root.get(), "stat", O_RDONLY)) {
+    return false;
+  }
+
+  ProcessContext output;
+
+  // clang-format off
+  auto succeeded = fs.enumFiles(
+    process_fdmap.get(),
+
+    [&](const std::string &name, bool directory) {
+      if (directory) {
+        return;
+      }
+
+      char* null_terminator{nullptr};
+      auto int_fd_value = std::strtoull(name.c_str(), &null_terminator, 10);
```

Comment:
Can you see if you can perform this conversion with one of our `tryTo<unsigned long long>` utilities?